cdap-watchdog/src/main/java/co/cask/cdap/logging/read/AvroFileReader.java

```diff
@@ -117,53 +117,46 @@ public void readLog(Location file, Filter logFilter, long fromTimeMs, long toTim
           return ImmutableList.of();
         }
 
-        GenericRecord datum;
         List<List<LogEvent>> logSegments = Lists.newArrayList();
         int count = 0;
 
         // Calculate skipLen based on fileLength
-        long skipLen = file.length() / 10;
+        long length = file.length();
+        LOG.trace("Got file Length {}", length);
+        long skipLen = length / 10;
         if (skipLen > DEFAULT_SKIP_LEN) {
           skipLen = DEFAULT_SKIP_LEN;
         } else if (skipLen <= 0) {
           skipLen = DEFAULT_SKIP_LEN;
         }
 
+        // For open file, start from file length and read till the actual eof
+        dataFileReader.sync(length);
+        long finalSync = dataFileReader.tell();
         List<LogEvent> logSegment = Lists.newArrayList();
+        logSegment = readToNextSync(logFilter, fromTimeMs, dataFileReader, -1, -1);
 
-        long lastSeekPos;
-        long seekPos = file.length();
-        while (seekPos > 0) {
-          lastSeekPos = seekPos;
-          seekPos = seekPos < skipLen ? 0 : seekPos - skipLen;
-          dataFileReader.sync(seekPos);
-
-          logSegment = logSegment.isEmpty() ? logSegment : Lists.<LogEvent>newArrayList();
-          // read all the elements in the current segment (seekPos up to lastSeekPos)
-          while (dataFileReader.hasNext() && !dataFileReader.pastSync(lastSeekPos)) {
-            datum = dataFileReader.next();
-
-            ILoggingEvent loggingEvent = LoggingEvent.decode(datum);
-
-            // Stop when reached fromTimeMs
-            if (loggingEvent.getTimeStamp() > fromTimeMs) {
-              break;
-            }
-
-            if (logFilter.match(loggingEvent)) {
-              ++count;
-              logSegment.add(new LogEvent(loggingEvent,
-                                          new LogOffset(LogOffset.INVALID_KAFKA_OFFSET, loggingEvent.getTimeStamp())));
-            }
-          }
+        long startPosition = finalSync;
+        long endPosition = startPosition;
+        long prevSync = finalSync;
+        long currentSync = finalSync;
 
+        while (startPosition > 0 && count < maxEvents) {
           if (!logSegment.isEmpty()) {
             logSegments.add(logSegment);
+            count = count + logSegment.size();
           }
-
-          if (count > maxEvents) {
-            break;
+          // Starting from endPosition, move backwards skipLen number of positions to find out largest sync point
+          // less than endPosition
+          while (endPosition == currentSync && startPosition > 0) {
+            startPosition = startPosition < skipLen ? 0 : startPosition - skipLen;
+            dataFileReader.sync(startPosition);
+            currentSync = dataFileReader.tell();
```

Comment:
After calling `dataFileReader.sync(startPosition)`, `dataFileReader.previousSync()` should also give the location of the position we sync-ed to, right? In that case we can directly use `endPosition = prevSync` later in the code and remove `currentSync`.

I think using `prevSync` is better than `dataFileReader.tell()`, because `tell()` gives the position of underlying stream, and depending on how the reader buffers that position might not exactly be the sync boundary.